server/bleep/src/agent/symbol.rs

```diff
@@ -0,0 +1,370 @@
+                        &document,
+                        &all_docs,
+                    )
+                }),
+        )
+        .await;
+
+        // filter references and definitions
+        let mut symbols = related_symbols
+            .into_iter()
+            .filter_map(Result::ok)
+            .zip(hoverable_ranges.into_iter().filter(|range| {
+                (range.start.byte >= chunk.start_byte) && (range.start.byte < chunk.end_byte)
+            }))
+            .map(|(token_info, range)| {
+                let filtered_token_info = token_info
+                    .into_iter()
+                    .filter(|file_symbols| file_symbols.file != chunk.path)
+                    .collect::<Vec<_>>();
+
+                Symbol {
+                    name: chunk.snippet.clone()[(range.start.byte - chunk.start_byte)
+                        ..(range.end.byte - chunk.start_byte)]
+                        .to_string(),
+                    related_symbols: filtered_token_info,
+                }
+            })
+            .filter(|metadata| {
+                (metadata.related_symbols.len() < MAX_REF_DEFS)
+                    && (!metadata.related_symbols.is_empty())
+            })
+            .collect::<Vec<_>>();
+
+        symbols.sort_by(|a, b| a.name.cmp(&b.name));
+        symbols.dedup_by(|a, b| a.name == b.name);
+
+        debug!("Attatched {} symbols", symbols.len());
+
+        Ok(ChunkWithSymbols {
+            chunk: chunk.clone(),
+            symbols,
+        })
+    }
+
+    pub async fn expand_symbol_into_chunks(&self, ref_def_metadata: Symbol) -> Vec<CodeChunk> {
+        const NUMBER_CHUNK_LINES: usize = 10;
+
+        let contents = ref_def_metadata
+            .related_symbols
+            .iter()
+            .map(|f_s| self.get_file_content(&f_s.file));
+
+        let contents = futures::future::join_all(contents)
+            .await
+            .into_iter()
+            .map(|content_document| content_document.unwrap().unwrap())
+            .collect::<Vec<_>>();
+
+        // each symbol may be in multiple files and have multiple occurences in each file
+        ref_def_metadata
+            .related_symbols
+            .iter()
+            .zip(contents.iter())
+            .flat_map(|(file_symbols, content)| {
+                let filename = file_symbols.file.clone();
+                let content = content.content.lines().collect::<Vec<_>>();
+
+                let n_lines = content.len();
+
+                file_symbols
+                    .data
+                    .iter()
+                    .map(|occurrence| {
+                        let chunk_content = content[occurrence.range.start.line
+                            ..(occurrence.range.end.line + NUMBER_CHUNK_LINES).min(n_lines)]
+                            .to_vec()
+                            .join("\n");
+                        CodeChunk {
+                            path: filename.clone(),
+                            alias: 0,
```

Comment:
We cannot set here because we will later truncate the chunks, so we want to avoid adding the alias/path in the prompt if the agent cannot see the chunk